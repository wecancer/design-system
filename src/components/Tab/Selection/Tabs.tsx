import React, {useEffect, useRef, useContext} from 'react'
import styled from 'styled-components'

import Context from './TabSection.context'

const Container = styled.header`
  display: flex;
  align-items: flex-start;
  margin-bottom: 1.25rem;
`

type Props = {
  children: React.ReactNode
}

const Tabs = ({children}: Props) => {
  const ref = useRef<HTMLElement>(null)
  const {setTabActiveId} = useContext(Context)

  useEffect(() => {
    const target = ref.current?.querySelectorAll('[data-tabtarget]')

    if (target?.length) {
      setTabActiveId(target[0].getAttribute('data-tabtarget'))
    }
  }, [setTabActiveId])

  return <Container ref={ref}>{children}</Container>
}

export default Tabs
