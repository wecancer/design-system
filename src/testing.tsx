import React from 'react'
import {render, RenderResult} from '@testing-library/react'
import WecancerProvider from './WecancerProvider'

type Props = {
  children: React.ReactNode
}

const AllTheProviders = ({children}: Props) => {
  return <WecancerProvider>{children}</WecancerProvider>
}

// eslint-disable-next-line
const appRender = (ui: React.ReactElement, options?: any): RenderResult =>
  render(ui, {wrapper: AllTheProviders, ...options})

// re-export everything
export * from '@testing-library/react'

// override render method
export {appRender as render}
