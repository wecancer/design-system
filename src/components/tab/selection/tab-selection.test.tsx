import { render } from '../../../testing'

import Tab from './tab'
import Tabs from './tabs'
import TabContext from '.'
import TabContent from './tab-content'

describe('<TabSection />', () => {
  it('should render snapshot correctly', () => {
    const { container } = render(
      <TabContext>
        <Tabs>
          <Tab target="0">First tab</Tab>
          <Tab target="1">Middle tab</Tab>
          <Tab target="2">Last tab</Tab>
        </Tabs>
        <TabContent id="0">This is my first tab</TabContent>
        <TabContent id="1">This is my middle tab</TabContent>
        <TabContent id="2">This is my last tab</TabContent>
      </TabContext>,
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the 5 items in the tab section', () => {
    const { getAllByRole } = render(
      <TabContext>
        <Tabs>
          <Tab target="0">First tab</Tab>
          <Tab target="1">Middle tab</Tab>
          <Tab target="2">Last tab</Tab>
        </Tabs>
        <TabContent id="0">This is my first tab</TabContent>
        <TabContent id="1">This is my middle tab</TabContent>
        <TabContent id="2">This is my last tab</TabContent>
      </TabContext>,
    )

    expect(getAllByRole('tab')).toHaveLength(3)
  })

  it('should have the label item', () => {
    const { getByText } = render(
      <TabContext>
        <Tabs>
          <Tab target="0">First tab</Tab>
          <Tab target="1">Middle tab</Tab>
          <Tab target="2">Last tab</Tab>
        </Tabs>
        <TabContent id="0">This is my first tab</TabContent>
        <TabContent id="1">This is my middle tab</TabContent>
        <TabContent id="2">This is my last tab</TabContent>
      </TabContext>,
    )

    const frist = getByText('First tab')

    expect(frist).toBeInstanceOf(HTMLElement)
  })

  it('should render the first item content by default', () => {
    const { getByText, queryByText } = render(
      <TabContext>
        <Tabs>
          <Tab target="0">First tab</Tab>
          <Tab target="1">Middle tab</Tab>
          <Tab target="2">Last tab</Tab>
        </Tabs>
        <TabContent id="0">This is my first tab</TabContent>
        <TabContent id="1">This is my middle tab</TabContent>
        <TabContent id="2">This is my last tab</TabContent>
      </TabContext>,
    )

    const fristContent = getByText('This is my first tab')
    const midleContent = queryByText('This is my middle tab')
    const lastContent = queryByText('This is my last tab')

    expect(fristContent).toBeInstanceOf(HTMLElement)
    expect(midleContent).toBeNull()
    expect(lastContent).toBeNull()
  })

  it('should active the Middle tab', () => {
    const { getByText } = render(
      <TabContext active="1">
        <Tabs>
          <Tab target="0">First tab</Tab>
          <Tab target="1">Middle tab</Tab>
          <Tab target="2">Last tab</Tab>
        </Tabs>
        <TabContent id="0">This is my first tab</TabContent>
        <TabContent id="1">This is my middle tab</TabContent>
        <TabContent id="2">This is my last tab</TabContent>
      </TabContext>,
    )

    const frist = getByText('This is my middle tab')

    expect(frist).toBeInstanceOf(HTMLElement)
  })
})
