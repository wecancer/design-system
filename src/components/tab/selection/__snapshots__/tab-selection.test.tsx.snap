// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TabSection /> should render snapshot correctly 1`] = `
<section
  class="sc-iBkjds liouGI"
>
  <header
    class="sc-gKXOVf fOIDDk"
  >
    <button
      aria-disabled="false"
      aria-selected="true"
      class="sc-bczRLJ bjlKKu sc-jSMfEi gJEbsC"
      data-tabtarget="0"
      role="tab"
      type="button"
    >
      First tab
    </button>
    <button
      aria-disabled="false"
      aria-selected="false"
      class="sc-bczRLJ bjlKKu sc-jSMfEi gJEbsC"
      data-tabtarget="1"
      role="tab"
      type="button"
    >
      Middle tab
    </button>
    <button
      aria-disabled="false"
      aria-selected="false"
      class="sc-bczRLJ bjlKKu sc-jSMfEi gJEbsC"
      data-tabtarget="2"
      role="tab"
      type="button"
    >
      Last tab
    </button>
  </header>
  This is my first tab
</section>
`;
